### PR TITLE
Emit events from the kernels service and gateway client

### PR DESCRIPTION
**Pull request by rajmusuku**
Event emitters using Jupyter events for GatewayClient and GatewayMappingKernelManager.

Introduced event schemas for GatewayClient and GatewayMappingKernelManager
Injected jupyter event logger from jupyter_events in the above classes.
Intercepted the Restful calls into Enterprise Gateway.
Emit event messages within Jupyter server.

### CODE REVIEW

**Comment by rajmusuku** [comment]
fixes #1253 #1253

File: jupyter_server/gateway/managers.py
Comment (kevin-bates) [inline jupyter_server/gateway/managers.py:79]: We may want to move emission from the "Mapping" Kernel Manager to the actual Kernel Manager, but this depends on how we decide to generalize this.  I thought, based on previous discussions we decided to emit from the specific manager instance which is why we introduced [`ServerKernelManager`](https://github.com/jupyter-server/jupyter_server/blob/6ad8006426df68583814c415848225a49f77373c/jupyter_server/services/kernels/kernelmanager.py#L706).

**Review by kevin-bates — COMMENTED** [review]
Hi @rajmusuku,

Thank for you opening this pull request, this is really great!  I do have a concern that it introduces a schema that is specific to _gateway_ kernel management when a general kernel management schema would be sufficient (and desired) and I think we should consider adopting a common schema for kernel management and from which `GatewayKernelManager` also emits events.  We introduced `ServerKernelManager` earlier for this specific reason for non-gateway kernels.

Since this PR is beating us to the mark for generic kernel management events, I think it would be good to at least generalize the schema you have here to be _the schema_ used for kernel management events in general.  We could then extend `ServerKernelManager` to emit the events from the same schema.

I'd be happy to help out with this if you like.  Thanks for getting this ball rolling!

**Comment by rajmusuku** [comment]
> 
@kevin-bates Thanks for your comment. It make sense to generalize the schema for Kernel service, in fact that was the initial thought to emit all the Kernel events. I would be more interested to capture failures than the regular transactions. Please let me know how can we work together on it.

**Comment by kevin-bates** [comment]
Hi @rajmusuku.  I think we should go ahead and generalize the kernel lifecycle event schema.  I think it makes sense to include a failure indication on the events but we can talk about.  We should review the generalized schema in the [server and kernel meeting](https://github.com/jupyter-server/team-compass#weekly-team-meetings) on Thursday.

If you'd like to get together prior to Thursday, please reach out to me via the email address associated with my GH profile.  If others would like to be involved in the initial discussion, please comment with where meeting invitations can be sent.

cc: @Zsailer

File: jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml]: Changed as suggested.

File: jupyter_server/event_schemas/gateway/client/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/client/v1.yaml]: Make sense.

File: jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml]: Done

File: jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml]: Changed description.

File: jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml]: Kept error as is since it will solve the problem of generic error handling of kernel manager.

Comment (kevin-bates) [reply jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml]: It's not clear to me if `error` should be an _action_ or the _status_/_result_ of a specific action.  For example, if we added a `status` or `result` field to each action and phrased the description of the action as an "attempt", then the status/result could be "OK" on success and the error string on failure (although a result code field might also be useful for more easily testing the result).

File: jupyter_server/event_schemas/kernels/v1.yaml
Comment (kevin-bates) [inline jupyter_server/event_schemas/kernels/v1.yaml:50]: Would it make sense to introduce a `kernel_server` or `server` property.  For Gateway-based kernels, this would be the host value parsed from the `gateway-url` and `localhost` (or the host parsed out of the initial url) for all others.

File: jupyter_server/event_schemas/gateway/client/v1.yaml
Comment (kevin-bates) [inline jupyter_server/event_schemas/gateway/client/v1.yaml:34]: Should this include the REST method and endpoint?

File: jupyter_server/event_schemas/kernels/v1.yaml
Comment (kevin-bates) [inline jupyter_server/event_schemas/kernels/v1.yaml:41]: We should probably rename this to `kernel_id` so that it's less ambiguous since "kernel" is commonly used to refer to the kernel name.  (Do we also need `kernel_name`?  We'd need to check that that information is available across the actions.)

File: jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml]: Similar to gateway client, we can add status and status_code properties.So based on the action result, we should be able to emit the success/error values.
https://github.com/jupyter-server/jupyter_server/blob/f50fb22564b5bbd0903ba5c28bbd395c9850db27/jupyter_server/event_schemas/gateway/client/v1.yaml#L12

Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/mapping_kernel_manager/v1.yaml]: Pushed latest changes.

File: jupyter_server/event_schemas/gateway/client/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/client/v1.yaml:10]: Added gateway_url property.

File: jupyter_server/event_schemas/gateway/client/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/client/v1.yaml:34]: _message_ includes exception details in case of errored scenario and a generic message from method in case of success..

File: jupyter_server/event_schemas/kernels/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/kernels/v1.yaml]: Added both, lets talk discuss the pattern.

File: jupyter_server/event_schemas/kernels/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/kernels/v1.yaml:79]: Adding server property make sense to me. Included the details.

File: jupyter_server/gateway/managers.py
Comment (rajmusuku) [reply jupyter_server/gateway/managers.py:75]: Initially, I started the implementation from KernelManger, however the KernelManager object is not exposed in serverapp, and therefore, I was not able to use the event logger in server extension that was registered in serverapp module.

File: jupyter_server/event_schemas/gateway/client/v1.yaml
Comment (kevin-bates) [reply jupyter_server/event_schemas/gateway/client/v1.yaml:34]: But can the event's consumer discern the endpoint that produced the event (success or failure)?  Just wondering if `endpoint` should be its own field.

**Comment by Zsailer** [comment]
Thanks for the ping, @kevin-bates. This is great! 

I'll spend some time today reviewing this more closing. I've been OOO sick the past week, so I'm just now catching up.

File: jupyter_server/event_schemas/kernels/v1.yaml
Comment (Zsailer) [reply jupyter_server/event_schemas/kernels/v1.yaml]: I think `kernel_name` should also be emitted.

**Comment by rajmusuku** [comment]
@kevin-bates @Zsailer Pushed changes as per our conversation. Please take a look, and let me know. Thanks!

**Comment by Zsailer** [comment]
Thanks @rajmusuku! 

I also push a new commit to your branch that begins emitting events from the kernel manager directly. If we agree with this approach, we can remove the emission from the MappingKernelManager.

We should also add some unit tests around these events. I'll start working on this...

**Comment by Zsailer** [comment]
Ok, I've added unit tests too 🚀

File: jupyter_server/event_schemas/kernel_actions/v1.yaml
Comment (kevin-bates) [inline jupyter_server/event_schemas/kernel_actions/v1.yaml:11]: I feel we should be consistent (the `gateway_client` schema uses `message`) and this should be expanded:
```suggestion
  - message
```

File: jupyter_server/gateway/managers.py
Comment (kevin-bates) [inline jupyter_server/gateway/managers.py:35]: This will need to match the schema property - which is currently `msg` but I've asked that it be expanded to `message`. - heads up.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (kevin-bates) [inline jupyter_server/services/kernels/kernelmanager.py:767]: Could we make this a constant?  Unlike the field names, this will likely change (per version) and seems worthy of a constant.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (kevin-bates) [inline jupyter_server/services/kernels/kernelmanager.py:773]: We should set the status and probably need a try/except clause here to appropriately log errors.

File: jupyter_server/event_schemas/kernel_actions/v1.yaml
Comment (Zsailer) [reply jupyter_server/event_schemas/kernel_actions/v1.yaml:11]: This is actually an (annoying and undocumented) limitation of jupyter_events. 

"message" is a special key that cannot be used in an event schema, because Python's underlying `logging` library mangles this key. See this chunk: https://github.com/jupyter/jupyter_events/blob/00d20622d413ea8b7ed950377f513d3a3db4d84c/jupyter_events/logger.py#L152-L159

We should add some logic to the jupyter_event's metaschema to exclude this key.

File: jupyter_server/gateway/managers.py
Comment (Zsailer) [reply jupyter_server/gateway/managers.py:35]: See this comment https://github.com/jupyter-server/jupyter_server/pull/1252#discussion_r1160785972

File: jupyter_server/services/kernels/kernelmanager.py
Comment (kevin-bates) [inline jupyter_server/services/kernels/kernelmanager.py:777]: I don't understand what this decorator buys us (but I find decorators confusing anyway).  Could you please explain its purpose?

File: jupyter_server/services/kernels/kernelmanager.py
Comment (Zsailer) [reply jupyter_server/services/kernels/kernelmanager.py:767]: Yeah. The only reason I didn't do this was, in the future, we might have multiple kernel schemas—e.g. a kernel state schema—so I was trying to make each call to the `emit(...)` method explicit and decoupled. That said, I don't think that's a strong enough argument. I can make a constant for this...

File: jupyter_server/services/kernels/kernelmanager.py
Comment (Zsailer) [reply jupyter_server/services/kernels/kernelmanager.py:777]: This ensures that the method "inherits" the same signature and docstring as the parent/overriden method. So the same arguments and keyword arguments are enforced. If the parent method's signature changes, this method automatically captures those changes as well. Further, the parent's docstring is inherited, which is handy as well.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (Zsailer) [reply jupyter_server/services/kernels/kernelmanager.py:773]: Status seems unique to gateway's kernel manager. What's the status in the (default) local case?

File: jupyter_server/event_schemas/gateway_client/v1.yaml
Comment (Zsailer) [inline jupyter_server/event_schemas/gateway_client/v1.yaml:33]: "message" is a (undocumented, unfortunately) disallowed key in jupyter_events. We probably need to change this to something like "msg".

File: jupyter_server/event_schemas/kernel_actions/v1.yaml
Comment (kevin-bates) [reply jupyter_server/event_schemas/kernel_actions/v1.yaml:11]: Interesting.  That's a drag since I'm sure `message` will be a common property of sorts.  Does the schema registration raise an exception when a "candidate schema" contains a `message` property?

Comment (Zsailer) [reply jupyter_server/event_schemas/kernel_actions/v1.yaml:11]: No it doesn't. We need to add this to the metaschema. 

We could probably add some logic to get around this issue in jupyter_events, I just haven't had the change to try this yet.

File: jupyter_server/gateway/gateway_client.py
Comment (Zsailer) [inline jupyter_server/gateway/gateway_client.py:109]: These must be `pathlib` objects to get the tests passing on Windows.

File: jupyter_server/gateway/managers.py
Comment (Zsailer) [inline jupyter_server/gateway/managers.py:52]: `os.path.join` is not right here. This is a URI, not a file path. Using `os.path.join` here would break things on Windows by inserting a backwards slash.

File: jupyter_server/gateway/managers.py
Comment (Zsailer) [inline jupyter_server/gateway/managers.py:62]: These must be `pathlib` objects to get the tests passing on Windows.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (kevin-bates) [reply jupyter_server/services/kernels/kernelmanager.py:777]: I see - thank you, that's helpful.  So, in theory, this _should_ be done for all subclasses?  I use `@overrides` for this in [gateway provisioners](https://github.com/jupyter-server/gateway_provisioners/blob/ec9b6952af006492d587e8e0ac0f0de09a44a805/gateway_provisioners/container.py#L120) which seems a bit cleaner and less confusing (IMHO).

Comment (Zsailer) [reply jupyter_server/services/kernels/kernelmanager.py:777]: Ooo I like that. It does add a (temporary) dependency, but it looks like there is an accepted PEP adding this to the standard lib: https://peps.python.org/pep-0698/

I'll go ahead an update.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (rajmusuku) [inline jupyter_server/services/kernels/kernelmanager.py:802]: This is good, however, it doesn't emit exceptions from super. You may need to catch exception, emit and then raise.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (rajmusuku) [inline jupyter_server/services/kernels/kernelmanager.py:765]: This is good, however, it doesn't emit exceptions from super. You may need to catch exception, emit and then raise.

**Comment by Zsailer** [comment]
BTW, I opened https://github.com/jupyter/jupyter_events/pull/72 to address the "message" field. We can wait until that PR is release and published before this PR is merged if we want to keep "message" as a valid field.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (kevin-bates) [reply jupyter_server/services/kernels/kernelmanager.py:802]: Since we have to duplicate the emit calls here and in Gateway's Kernel Manager, I think its important that the content for the events be identical - same data and message formats - otherwise event-processing applications will have a frustrating time having to deal with two different "fingerprints".

File: jupyter_server/event_schemas/kernel_actions/v1.yaml
Comment (kevin-bates) [reply jupyter_server/event_schemas/kernel_actions/v1.yaml:11]: If the events PR becomes too involved we could use a different name.  Some suggestions: `text`, `event_message`, `body`, `content`, ...?  I kinda like `content`.

Comment (rajmusuku) [reply jupyter_server/event_schemas/kernel_actions/v1.yaml:11]: I think, `msg` should be okay compare to other key words. `body` make sense to me but it confuses us with REST `body` which is little different. Following are a few more options.
`comment`
`notification`

File: jupyter_server/gateway/managers.py
Comment (rajmusuku) [reply jupyter_server/gateway/managers.py:76]: Agree. Is it okay if I remove in a separate PR ? Let me test the current behavior with an extension and circle back otherwise.

File: jupyter_server/gateway/managers.py
Comment (rajmusuku) [reply jupyter_server/gateway/managers.py]: I changed it to `msg` to be in sync with kernel_actions for now.

File: jupyter_server/event_schemas/gateway_client/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway_client/v1.yaml]: Done, changed it to `msg` to be in sync with `kernel_actions`

File: jupyter_server/services/kernels/kernelmanager.py
Comment (kevin-bates) [reply jupyter_server/services/kernels/kernelmanager.py:773]: Status will typically be success, but requests can fail - so the status should reflect the result of the action.

**Comment by kevin-bates** [comment]
@Zsailer - I'm loving that last commit (adding decorator `emit_kernel_action_event`)!  Thank you!

**Comment by Zsailer** [comment]
Thanks @kevin-bates! 

Are we okay with moving the kernel event emission into the GatewayKernelManager, similar to how we do in the default ServerKernelManager? This enables us to use the `emit_kernel_action_event` there as well. 

I've added a unit test to show that if the underlying exception inside a kernel method is a tornado `HTTPError`, it will try to set the status_code. Does this cover your use-case @rajmusuku?

File: jupyter_server/services/kernels/kernelmanager.py
Comment (Zsailer) [reply jupyter_server/services/kernels/kernelmanager.py]: Fixed in my latest commits.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (Zsailer) [reply jupyter_server/services/kernels/kernelmanager.py]: The new decorator I created, `emit_kernel_action_event`, eliminates the need to maintain duplicates of these events.

File: jupyter_server/gateway/managers.py
Comment (Zsailer) [reply jupyter_server/gateway/managers.py]: Shoot—I missed this comment. Sorry @rajmusuku. 

I went ahead and moved the event emission down into the individual kernel manager. Let me know what you think.

File: jupyter_server/services/kernels/kernelmanager.py
Comment (Zsailer) [reply jupyter_server/services/kernels/kernelmanager.py]: :+1: added "status" to all emitted events.

**Comment by kevin-bates** [comment]
> Are we okay with moving the kernel event emission into the GatewayKernelManager, similar to how we do in the default ServerKernelManager? This enables us to use the emit_kernel_action_event there as well.

Absolutely!  Thank you.  This also addresses the message consistency issue - so this is really great!

**Comment by rajmusuku** [comment]
> 

@Zsailer Perfect, it covers my scenario. Also, let's pass the exception message in `msg` property, if that make sense.

**Comment by rajmusuku** [comment]
> Thanks @kevin-bates!
> 
> Are we okay with moving the kernel event emission into the GatewayKernelManager, similar to how we do in the default ServerKernelManager? This enables us to use the `emit_kernel_action_event` there as well.
> 
> I've added a unit test to show that if the underlying exception inside a kernel method is a tornado `HTTPError`, it will try to set the status_code. Does this cover your use-case @rajmusuku?

Yes, it covers my scenario and in general use cases too. Thanks!!

**Comment by Zsailer** [comment]
This branch diverged a bit from main and was causing a nasty rebase. Looks like you were might have initially branched off from a non-`main` branch? This caused many of the commit SHAs to be different between your branch and main. 

Since we squash commits on merge anyways, I went ahead and manually squashed all of these commits into a single, new commit. I've manually set you, @rajmusuku, as the author and myself as co-author. Annoyingly, Github still shows my picture in the UI above, but if you look at the details of the commit, you're listed as the main author. Sorry for the all git gymnastics. 

I think this is ready to go. @kevin-bates do you mind giving this one more look before merging? 

Tests are failing for something unrelated (looks like codecov pypi issues).

File: jupyter_server/gateway/managers.py
Comment (kevin-bates) [inline jupyter_server/gateway/managers.py:297]: This seems redundant since we're logging the same information prior to the gateway request.  Given this occurs at a (lab-induced) polling interval of every 10 seconds and an exception will be raised to indicate a failure, I'd prefer we remove this statement and go with a "no news is good news" strategy here and eliminate this extra noise.  Does that sound okay?
```suggestion
```

File: jupyter_server/gateway/managers.py
Comment (kevin-bates) [inline jupyter_server/gateway/managers.py:425]: I think we should derive `GatewayKernelManager` from `ServerKernelManager` (this should have been done when the latter got introduced a while back actually - my bad), then let inheritance bring in these attributes since these should be identical to those defined in `ServerKernelManager`.  
```suggestion
```

**Comment by kevin-bates** [comment]
Here's the info regarding the `codecov` issues (it was removed from pypi earlier today): https://community.codecov.com/t/codecov-yanked-from-pypi-all-versions/4259/1

File: jupyter_server/gateway/managers.py
Comment (Zsailer) [reply jupyter_server/gateway/managers.py]: :+1: Removed.

File: jupyter_server/gateway/managers.py
Comment (Zsailer) [reply jupyter_server/gateway/managers.py]: Sounds good. I've updated it. 

I wasn't sure if we wanted to do this, since GatewayKernelManager inherited from `AsyncKernelManager` while ServerKernelManager inherits from `AsyncIOLoopKernelManager`. Will this be an issue?

**Comment by Zsailer** [comment]
Thank for tracking down the codecov issue, @kevin-bates! 

We can address the that mess in a separate PR. 😅

File: jupyter_server/gateway/managers.py
Comment (kevin-bates) [reply jupyter_server/gateway/managers.py]: Thanks.  Yeah, I suspect `GatewayKernelManager` was incorrect all along and should have derived from `AsyncIOLoopKernelManager` from the get-go.  It really doesn't call `super()` anyway, essentially deriving method signatures (and attributes) and overriding those where necessary (e.g., `_default_cache_ports()`).

(Looking now, GKM does call `super().__init__()`, but AILKM does define that method, while AKM does, so we're not introducing any new code there.  We should be good.)

**Review by kevin-bates — APPROVED** [review]
This is great stuff - thank you @rajmusuku and @Zsailer!

@rajmusuku - I'm hoping you can take this for a spin in your environment where you actually have an event listener configured.

**Comment by rajmusuku** [comment]
> 

Thank you @Zsailer, @kevin-bates for your input and help. Yes, will verify it tomorrow and let you know. However, we can merge this PR.

**Comment by kevin-bates** [comment]
>> I'm hoping you can take this for a spin in your environment where you actually have an event listener configured.

> However, we can merge this PR.

Hi @rajmusuku - I made my comment because there aren't gateway-specific tests for confirming the event's emission from `GatewayKernelManager`.  I apologize for not being more clear about that aspect but was hesitant to ask.  I can try to address this today

**Comment by kevin-bates** [comment]
Welp, it turns out [this comment](https://github.com/jupyter-server/jupyter_server/pull/1252#discussion_r1164804027) was incorrect, and derivation from `AsyncIOLoopKernelManager` is (somehow) problematic.  I'd like to better understand this before duplicating the decorator.  When deriving from AILKM, we are unable to obtain an instance [here](https://github.com/jupyter-server/jupyter_server/pull/1252/files#diff-2db15f098ae2b79c816989bdf615b0d1f9b2c728e3d321acafa39f0de5233643R81):
```python
km = self.kernel_manager_factory(parent=self, log=self.log)
```

**Comment by kevin-bates** [comment]
> Welp, it turns out https://github.com/jupyter-server/jupyter_server/pull/1252#discussion_r1164804027 was incorrect, and derivation from AsyncIOLoopKernelManager is (somehow) problematic. I'd like to better understand this before duplicating the decorator.

It (actually) turns out to have been an incompatibility between `GatewayKernelManager` and `ServerKernelManager` because the latter (correctly) types `last_activity` and the former was setting its value to `None`.  I've gone ahead and set initial values (to `execution_state` as well) and pushed the correction.  The gateway tests should pass.

I'd still like to get _some_ event testing for gateway completed and will focus on that shortly.  Sorry for the delays!

File: jupyter_server/services/kernels/kernelmanager.py
Comment (kevin-bates) [inline jupyter_server/services/kernels/kernelmanager.py:813]: @Zsailer - because the event_logger trait is on the `ServerKernelManager` and not the `AsyncMappingKernelManager`, do we need the _double-hop_ of parent to access the configurable on `ServerApp` here?  I.e.,:
```suggestion
        if self.parent is not None and self.parent.parent is not None and hasattr(self.parent.parent, "event_logger"):
```

File: tests/services/kernels/test_events.py
Comment (kevin-bates) [inline tests/services/kernels/test_events.py:15]: @Zsailer. I'm looking into adding event handling to the gateway tests and find this block of lines a little confusing and I'm not sure if this is a practical test.

If the _handler_ for an event is meant for an event's _consumer_ and the `ServerKernelManager` instance does not get instantiated until a kernel is started, how would a consumer be able to set their handler on _that_ instance?  It seems like handler registration can only take place on "singleton instances" - like `ContentManager` and `MappingKernelManager`, etc. where the instance is created via `init_configurables()`.

Hmm - I think I need to go RTD to understand the distinction between handlers and listeners.

**Comment by rajmusuku** [comment]
@Zsailer @kevin-bates Just tested with my extension, however, the event listener is not catching the emitted events. Will do some investigation and let you know.

**Comment by rajmusuku** [comment]
> @Zsailer @kevin-bates Just tested with my extension, however, the event listener is not catching the emitted events. Will do some investigation and let you know.

@Zsailer @kevin-bates Let's merge this PR as we are good with all the changes. I will work on more testing with my extension. Please let me know if you have any questions.

File: jupyter_server/event_schemas/gateway/client/v1.yaml
Comment (rajmusuku) [reply jupyter_server/event_schemas/gateway/client/v1.yaml]: We added `gateway_url` to handle this specific scenario. Thanks!

**Comment by blink1073** [comment]
I removed those checks from the main branch, new PRs should have them.

**Comment by Zsailer** [comment]
> Just tested with my extension, however, the event listener is not catching the emitted events. Will do some investigation and let you know.

I think we should add documentation about events in jupyter_server before merging. I'll work on this today.

**Comment by kevin-bates** [comment]
@Zsailer - I'm good to go here, are you good to go here? :smile:

**Comment by Zsailer** [comment]
> I'm good to go here, are you good to go here?

Let me open a separate PR showing the documentation I propose. 

We don't have to block on that other PR being merged, but lets at least wait until I open that documentation PR (give me a hour or two).

**Comment by Zsailer** [comment]
Ok, I opened #1260 to begin documenting the events emitted by Jupyter Server. I'll keep that in draft state until I'm able to get the auto-generation of event schema documentation working. 

In the meantime, I think this is ready to go. Merging!

**Pull request merged**
